Feature/questions answers/markdown

filip-gregurevic opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Students, tutors and instructors can only use plain text to write questions and answers limiting their ability to describe the problem or solution precisely.
The buttons of the Markdown editor take too much space and don't look good on smaller screens.

### Description
<!-- Describe your changes in detail -->
I added the Markdown editor to the questions and answers component.
I changed the design of the buttons in the markdown editor UI to be smaller in order to fit better on smaller screens / spaces.

**NOTE: If you encounter any bugs with the markdown editor please let me know. I will be further working on it in a separate PR, in which I will also address the functionality.**
(Also if you have any wishes or suggestions for the Markdown editor)

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
(1a. Enable questions and answers on a course if deactivated) 
2. Navigate to a lecture or exercise
3. Markdown should now be available for questions and answers
4. View any other Markdown editor and check if buttons are displayed correctly

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
Markdown editor for question (with smaller buttons):
<img width="419" alt="Bildschirmfoto 2020-04-20 um 19 04 16" src="https://user-images.githubusercontent.com/51987021/79779754-2547e800-833b-11ea-9bbb-01e980057979.png">

Editor after changes:
<img width="417" alt="Bildschirmfoto 2020-04-20 um 22 59 45" src="https://user-images.githubusercontent.com/51987021/79799455-c3977600-835a-11ea-8f91-0ec5ce558083.png">

krusche commented:
Nice work. Looks much better now 👍 I would really love to ship this in the release 4.0.1 before all students start to use Artemis 🥇 

Can we please get rid of the 3rd row above the text editor in your screenshot? I think 2 rows are really the absolut maximum here. 
I think we can move the fullscreen button to the first row (for all Markdown editors who want to have this enabled) with right alignment. It makes more sense here, because it's not really a command, but a view control.

For questions and answers specifically, we should remove the "Formula" button, because this is rarely used to format questions.

Then, we would already have 2 rows ;-)

filip-gregurevic commented:
> Nice work. Looks much better now 👍 I would really love to ship this in the release 4.0.1 before all students start to use Artemis 🥇
> 
> Can we please get rid of the 3rd row above the text editor in your screenshot? I think 2 rows are really the absolut maximum here.
> I think we can move the fullscreen button to the first row (for all Markdown editors who want to have this enabled) with right alignment. It makes more sense here, because it's not really a command, but a view control.
> 
> For questions and answers specifically, we should remove the "Formula" button, because this is rarely used to format questions.
> 
> Then, we would already have 2 rows ;-)

@krusche thanks for the feedback!
I removed the formula button now and added the option to remove the fullscreen button, which I also deactivated here. It is enabled by default so that it doesn't interfere with the editors used elsewhere.
<img width="417" alt="Bildschirmfoto 2020-04-20 um 22 59 45" src="https://user-images.githubusercontent.com/51987021/79799944-a1eabe80-835b-11ea-9ef0-b811ef6d04dc.png">

This makes it only two rows for now.
The goal for the next update (PR) is to make it only one row, but that will take a little bit more of refactoring.

krusche commented:
@filip-gregurevic Could you please resolve the conflict and fix the typescript tests? thanks!

filip-gregurevic commented:
@krusche all fixed now 👍

HanyaElhashemy commented:
And as for the answer editor:

- The full screen option is missing (is this intentionally)
- The quote command is not working as expected but this is a known issue for all markdown editors, so not really related with this PR
![image](https://user-images.githubusercontent.com/50635005/79852264-b87a2f80-83c6-11ea-92da-3b11b3822404.png)

filip-gregurevic commented:
> Great work, I really like the new feature!
> I noticed a missing functionality in the markdown editor that you maybe could address in a follow-up PR:
> 
> * The heading (Styling) command is not working as expected (check preview tab after using a Heading2 for example)
> 
> Would it be possible to add the fullscreen command in the same line as the other commands bar. By that you will save some wasted space and maybe make the initial text field a bit bigger
> ![image](https://user-images.githubusercontent.com/50635005/79851603-d72bf680-83c5-11ea-9b34-0ec118f68e33.png)

@HanyaElhashemy thanks for the feedback!
I removed the fullscreen command intentionally on the questions and answers. I just forgot to remove it for the new questions, but it should be fixed with the latest commit. 
I will refactor the whole layout of the Markdown editor in another PR and I will check all functionalities / commands. I noted the ones you mentioned, thanks :)

madwau commented on src/main/webapp/app/shared/markdown-editor/markdown-editor.component.html:
```diff
@@ -27,42 +27,48 @@
+                    <!-- domain commands -->
+                    <div class="markdown-editor__commands-domain" *ngIf="domainCommands && domainCommands.length != 0">
+                        <ng-container *ngFor="let command of domainCommands | typeCheck: DomainTagCommand">
+                            <div *ngIf="command.displayCommandButton === true" class="btn btn-sm" (click)="command.execute()">
```
Can this be simplified to `command.displayCommandButton`?

filip-gregurevic replied:
Good catch! I didn't notice it, while rearranging the commands. Fixed with https://github.com/ls1intum/Artemis/pull/1325/commits/cb8b3bbfc0d8d43792ef8f1033ee0fbfcb125008

madwau commented on src/main/webapp/app/overview/student-questions/student-question-row.component.html:
```diff
@@ -49,12 +52,19 @@ <h6 class="text-black-50 question-subtitle">{{ studentQuestion.author.name }} -
+                            [(markdown)]="studentQuestionText"
+                            [editorMode]="EditorMode.LATEX"
+                            [enableResize]="true"
+                            [metaCommands]="[]"
```
Could `[]` be a default value for the Input that doesn't need to be passed in explicitly?

filip-gregurevic replied:
I needed to do it that way for now in order to keep the fullscreen option in every other usage of the markdown editor and disable it here, but I will change this when refactoring the whole editor in a follow-up PR.

madwau commented on src/main/webapp/app/overview/student-questions/student-question-row.component.html:
```diff
@@ -63,53 +73,102 @@ <h6 class="text-black-50 question-subtitle">{{ studentQuestion.author.name }} -
-            <div class="col" [ngbTooltip]="'artemisApp.courseOverview.exerciseDetails.faq.approvedLabel' | translate">
+            <div class="col">
                 <div *ngIf="isAnswerMode && selectedQuestionAnswer && approvedQuestionAnswer && selectedQuestionAnswer.id === approvedQuestionAnswer.id">
                     <div class="col-12">
```
A bootstrap `.col` should always be a child of a `.row`.

filip-gregurevic replied:
fixed with https://github.com/ls1intum/Artemis/pull/1325/commits/cb8b3bbfc0d8d43792ef8f1033ee0fbfcb125008

madwau commented on src/main/webapp/app/overview/student-questions/student-question-row.component.html:
```diff
@@ -135,13 +195,52 @@ <h6 class="text-black-50 question-subtitle">{{ studentQuestion.author.name }} -
                     </div>
                 </div>
             </div>
+            <!-- unapproved answers -->
```
The code below looks very similar to the one for the approved answers. There also seems to be a lot of shared template code between the view and edit state (see screenshot below).

I think it would make sense to avoid this duplication by creating small new components whenever reusable parts like these manifest themselves.

Code duplicate:
![OnPaste 20200421-121427](https://user-images.githubusercontent.com/7109225/79854428-b6fe3680-83c9-11ea-9bc7-04794361b381.png)

filip-gregurevic replied:
I will address the code duplication (or at least very similar code) in a follow-up PR, in which I change the layout of the questions and answers (and also refactor them into smaller components).
@HanyaElhashemy

madwau replied:
Great, thanks!

HanyaElhashemy commented:
@filip-gregurevic The markdown layout looks good now, thanks for addressing my comments ^^
One last thing: could you please try to reduce the duplicate html code and since question and answer have the same markdown editor layout you could define a component for the layout and use it for both the question and answer in Exercise as well as in Lecture

TobiasPr commented:
I also did some testing. My findings are more related to your next pull request:
Fore me coloring of text did not work 

testing:
`# Heading 1
**test**
*test*
<ins>test</ins>
> test
```language test```
[test](http://www.google.de)
1. test
2. test
- test
- test
## Heading 2
# Heading 1
### Heading 3
<span class="red"> test red </span>
$$ e^{\frac{1}{4} y^2} $$`

filip-gregurevic commented:
> I also did some testing. My findings are more related to your next pull request:
> Fore me coloring of text did not work and also the first heading. Sometimes newlines \n were skipped
> 
> testing:
> 
> ```
> **test**
> *test*
> <ins>test</ins>
> > test
> ```language test```
> [test](http://www.google.de)
> 1. test
> 2. test
> - test
> - test
> ## Heading 2
> # Heading 1
> ### Heading 3
> <span class="red"> test red </span>
> $$ e^{\frac{1}{4} y^2} $$
> ```
> 
> What I would also like to see if the answer field would expand a bit and not only display one line. One could argure thats what the full screen feature is for, but I think most answers will be bigger than one line and ImO that would be nice to have
> ![image](https://user-images.githubusercontent.com/22955066/79857150-a0f27500-83cd-11ea-8c04-dfae2dbc8bf9.png)

@TobiasPr thanks! I noted the bugs.
And there's the option to resize the answer/ question window below the text area for longer texts, but I will also set a bigger default hight in the new layout

krusche commented:
@filip-gregurevic I also think the initial size of the text area for questions and answers should be larger. Can you set it to 100px (min) height?

filip-gregurevic commented:
> @filip-gregurevic I also think the initial size of the text area for questions and answers should be larger. Can you set it to 100px (min) height?

@krusche done with the latest commit.

madwau commented:
Probably unrelated to this PR but I just noticed that the scrollbars are always shown for the preview which looks a bit odd. If it's easy to set them to `auto`, that would be really nice. If not, also not a big deal.

![OnPaste 20200421-165317](https://user-images.githubusercontent.com/7109225/79880675-9eeddd80-83f0-11ea-901b-72a63027fdb4.png)

filip-gregurevic commented on src/main/webapp/app/shared/markdown-editor/markdown-editor.component.html:
```diff
@@ -27,42 +27,48 @@
                     <!--color picker command -->
                     <div class="btn-group col-xs-6">
-                        <div class="color-preview form-control btn btn-outline-secondary" (click)="openColorSelector($event)">
+                        <div class="color-preview btn btn-sm" [ngStyle]="{ borderColor: selectedColor }" (click)="openColorSelector($event)">
```
@balazs-czopf makes total sense, I'll include this in the next PR. I guess it must have slipped back in when resolving the merge conflicts. Thanks! :)

krusche replied:
Good catch, I'll remove this directly on develop then.
